### PR TITLE
feat: Support for URL in the filter rules file just like maven FindBugs plugin #253

### PR DESCRIPTION
**Pull request by msiemczyk**
Use Property so that the either URL or File can be given. If the URL is given then the rules file is downloaded to the temp location first and plugin uses it from there. Also added two functional tests to cover the new use cases.

### CODE REVIEW

File: src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy [hunk 1]
```diff
@@ -40,8 +40,8 @@ import org.gradle.api.provider.Property;
  * &nbsp;&nbsp;&nbsp;&nbsp;visitors = [ 'FindSqlInjection', 'SwitchFallthrough' ]<br>
  * &nbsp;&nbsp;&nbsp;&nbsp;omitVisitors = [ 'FindNonShortCircuit' ]<br>
  * &nbsp;&nbsp;&nbsp;&nbsp;reportsDir = file("$buildDir/reports/spotbugs")<br>
- * &nbsp;&nbsp;&nbsp;&nbsp;includeFilter = file('spotbugs-include.xml')<br>
- * &nbsp;&nbsp;&nbsp;&nbsp;excludeFilter = file('spotbugs-exclude.xml')<br>
+ * &nbsp;&nbsp;&nbsp;&nbsp;includeFilter = file('spotbugs-include.xml') or new URL('http://sonar?findbugs-include.xml')<br>
+ * &nbsp;&nbsp;&nbsp;&nbsp;excludeFilter = file('spotbugs-exclude.xml') or new URL('http://sonar?findbugs-exclude.xml')<br>
```
Comment (KengoTODA) [inline src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy:44]: better to use valid URL like `https://example.com/path/to/findbugs-include.xml`

File: src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy [hunk 2]
```diff
@@ -40,8 +40,8 @@ import org.gradle.api.provider.Property;
  * &nbsp;&nbsp;&nbsp;&nbsp;visitors = [ 'FindSqlInjection', 'SwitchFallthrough' ]<br>
  * &nbsp;&nbsp;&nbsp;&nbsp;omitVisitors = [ 'FindNonShortCircuit' ]<br>
  * &nbsp;&nbsp;&nbsp;&nbsp;reportsDir = file("$buildDir/reports/spotbugs")<br>
- * &nbsp;&nbsp;&nbsp;&nbsp;includeFilter = file('spotbugs-include.xml')<br>
- * &nbsp;&nbsp;&nbsp;&nbsp;excludeFilter = file('spotbugs-exclude.xml')<br>
+ * &nbsp;&nbsp;&nbsp;&nbsp;includeFilter = file('spotbugs-include.xml') or new URL('http://sonar?findbugs-include.xml')<br>
```
Comment (KengoTODA) [inline src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy:43]: could you use comments to explain the usage of URL, or this snippet becomes invalid groovy script so user cannot copy-and-paste.

File: src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy [hunk 3]
```diff
@@ -98,7 +99,8 @@ class SpotBugsExtension {
      * <p>See also <a href="https://spotbugs.readthedocs.io/en/stable/filter.html">SpotBugs Manual about Filter file</a>.</p>
      */
     @NonNull
-    final RegularFileProperty includeFilter;
+    final Property<Object> includeFilter;
```
Comment (KengoTODA) [inline src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy:102]: this field is public API, so it's better to keep it compatible. I suggest you to add a new field, and link `includeFilter` prop to it via a convention. like:

```groovy
includeFilter.convention(() -> downloadFrom(includeFilterUrl))
```

or, it's better to break backward compatibility, make next release as a major update (v5.0.0). Refer https://www.conventionalcommits.org/en/v1.0.0/ to learn how to release `BREAKING CHANGE`.

File: src/main/groovy/com/github/spotbugs/snom/SpotBugsTask.groovy [hunk 4]
```diff
@@ -416,4 +429,28 @@ class SpotBugsTask extends DefaultTask implements VerificationTask {
...
+                tempFile.deleteOnExit()
+
+                tempFile.withOutputStream { out ->
+                    out << filter.openStream()
+                }
+
+                filterFileProperty.set(tempFile)
+            }
+            else {
+                throw new IllegalArgumentException("The '$filterName' property must be a File or URL.")
```
Comment (KengoTODA) [inline src/main/groovy/com/github/spotbugs/snom/SpotBugsTask.groovy:452]: This exception is thrown to Gradle core (our code does not catch this exception), then I suggest using a subclass of `GradleException` like [InvalidUserDataException](https://docs.gradle.org/current/javadoc/org/gradle/api/InvalidUserDataException.html).

**Comment by msiemczyk** [comment]
@KengoTODA Thank you for your feedback, but I already spent too much time on this and I don't really have more time to make the changes you requested. Also we wouldn't be able to wait to version 5.0 in order to use this functionality; we are moving our build system now and I need to restore all out features from maven. For these reasons I've already built and published custom version of this plugin to our own maven and will just use that until the public version of this plugins supports URL for filter files.

Of course you are always welcome to make these changes yourself to enhance the plugin for others.

File: src/main/groovy/com/github/spotbugs/snom/SpotBugsTask.groovy (hunk 4 above)
Comment (msiemczyk) [reply src/main/groovy/com/github/spotbugs/snom/SpotBugsTask.groovy:452]: What difference will that make?

File: src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy (hunk 3 above)
Comment (msiemczyk) [reply src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy:102]: I don't understand your comment by keeping it compatible. The property is backwards compatible the way I wrote the code. All you loose is strongly typed RegularFileProperty, but I didn't see any other way for this one setting to accept to different types.

If you add new property, includeFilterUrl, then you will also need extra user input validation to make sure both things are not specified at the same time. Also I don't really understand that "convention()" thing and Gradle has rather poor documentation.

File: src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy (hunk 2 above)
Comment (msiemczyk) [reply src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy:43]: So you would want the '...or new URL' part removed from here and added below instead?

Comment (msiemczyk) [reply src/main/groovy/com/github/spotbugs/snom/SpotBugsExtension.groovy:43]: So you would want the '...or new URL' part removed from here and added below instead?

**Pull request closed without merging**